clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java

```diff
@@ -507,12 +503,44 @@ protected CreateServiceRequest makeServiceRequest(
             .withServiceRegistries(serviceRegistries)
             .withDeploymentConfiguration(deploymentConfiguration);
 
+    List<Tag> taskDefTags = new LinkedList<>();
     if (description.getTags() != null && !description.getTags().isEmpty()) {
-      Collection<Tag> taskDefTags = new LinkedList<>();
       for (Map.Entry<String, String> entry : description.getTags().entrySet()) {
         taskDefTags.add(new Tag().withKey(entry.getKey()).withValue(entry.getValue()));
       }
-      request.withTags(taskDefTags).withEnableECSManagedTags(true).withPropagateTags("SERVICE");
+    }
+
+    // Apply moniker strategy which may add tags
+    namer.applyMoniker(
+        new EcsResource() {
+          @Override
+          public String getName() {
+            return request.getServiceName();
+          }
+
+          // Used by Frigga when moniker support is disabled
+          public void setName(String name) {
+            request.setServiceName(name);
+          }
+
+          @Override
+          public List<Tag> getResourceTags() {
+            return taskDefTags;
+          }
+        },
+        newServerGroupName.getMoniker());
+
+    // Only add tags if they're set as it's an optional feature for ECS
+    if (taggingEnabled) {
+      request
+          .withTags(taskDefTags)
+          .withEnableECSManagedTags(true)
+          .withPropagateTags(PropagateTags.SERVICE.toString());
+    } else {
+      if (!taskDefTags.isEmpty()) {
+        throw new IllegalArgumentException(
+            "Tagging is not enabled on ECS account " + description.getAccount() + ".");
+      }
```

Comment:
It might make sense to do this prior to creating the request (fail fast).

Reply:
It has to be checked after the call to `namer.applyMoniker()` as that may have added tags even if the original operation description didn't include tags.

I don't think there's any large cost to checking after the `CreateServiceRequest` is created. The only thing I can see in the code that might be affected is the call to describe load balancers to attach. That should be moved after the tag setting check.